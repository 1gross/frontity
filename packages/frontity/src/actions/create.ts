import { resolve } from "path";
import ora from "ora";
import { prompt, Question } from "inquirer";
import create from "../functions/create";
import errorLogger from "../utils/error";
import { EventEmitter } from "events";
import { Options } from "../functions/create/types";

export default async (name: string, { typescript, useCwd }) => {
  const options: Options = {};

  if (!name) {
    const questions: Question[] = [
      {
        name: "name",
        type: "input",
        message: "Enter a name for the project:",
        default: "my-frontity-project"
      },
      {
        name: "theme",
        type: "input",
        message: "Enter a starter theme to clone:",
        default: "@frontity/mars-theme"
      },
      {
        name: "packages",
        type: "input",
        message: "Enter a list of Frontity packages to install:",
        default: "frontity, @frontity/file-settings",
        filter: (input: string) => input.split(/[\s,]+/)
      }
    ];

    const answers = await prompt(questions);
    options.name = answers.name;
    options.packages = answers.packages;
    options.theme = answers.theme;
    console.log();
  } else {
    options.name = name;
  }

  options.typescript = typescript;
  options.path = useCwd ? process.cwd() : resolve(process.cwd(), options.name);
  options.emitter = new EventEmitter();

  options.emitter.on("create", (message, action) => {
    if (action) ora.promise(action, message);
    else console.log(message);
  });

  options.emitter.on("error", error => {
    errorLogger(error);
  });

  await create(options);
};
