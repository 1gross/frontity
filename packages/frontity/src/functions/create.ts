import { EOL } from "os";
import { resolve } from "path";
import {
  ensureDir,
  readdir as readDir,
  readFile,
  writeFile,
  remove
} from "fs-extra";
import { exec } from "shelljs";
import { extract } from "tar";
import chalk from "chalk";
import { mergeRight } from "ramda";
import { CreateOptions, PackageJson } from "../types";

let dirExisted: boolean = false;

const defaultOptions: CreateOptions = {
  path: process.cwd(),
  typescript: false,
  packages: ["frontity", "@frontity/file-settings"],
  theme: "@frontity/mars-theme"
};

export const normalizeName = (name: string): string =>
  name.replace(/[\s_-]+/g, "-").toLowerCase();

// This function normalizes and validates options.
const normalizeOptions = (passedOptions: CreateOptions): CreateOptions => {
  const options = mergeRight(defaultOptions, passedOptions);

  // Normalize and valiidate `name` option.
  options.name = normalizeName(options.name);
  const nameConventionMatch = /^(?:@[\w-]+\/)?[\w-]+$/;
  if (!nameConventionMatch.test(options.name))
    throw new Error(
      "The name of the package is not valid. Please enter a valid one (only letters and dashes)."
    );

  return options;
};

// This function ensures that the directory exists and is empty.
const ensureAppDir = async ({ path }: CreateOptions) => {
  await ensureDir(path);
  process.chdir(path);
  const dirContent = await readDir("./");

  // If the directory is not empty check if it's an empty repository.
  if (dirContent.length) {
    const allowedContent = ["readme.md", "license", ".gitignore", ".git"];
    const notAllowedContent = dirContent.filter(
      content => !allowedContent.includes(content.toLowerCase())
    );
    // If it's not an empty repository.
    if (notAllowedContent.length) {
      throw new Error("The directory passed to `create` function is not empty");
    } else {
      dirExisted = true;
    }
  }
};

// This function creates a `package.json` file.
const createPackageJson = async ({ name }: CreateOptions) => {
  const packageJson: PackageJson = {
    name,
    version: "0.1.0",
    description: "Frontity project",
    keywords: ["frontity"],
    scripts: {
      dev: "frontity dev",
      build: "frontity build",
      serve: "frontity serve"
    }
  };
  const fileName = "package.json";
  const fileData = `${JSON.stringify(packageJson, null, 2)}${EOL}`;
  await writeFile(fileName, fileData);
};

// This function creates a `frontity.settings` file.
const createFrontitySettings = async (
  fileExtension: string,
  { name, packages }: CreateOptions
) => {
  const frontitySettings = { name, packages };
  const fileTemplate = await readFile(
    resolve(__dirname, `../templates/settings-${fileExtension}-template`),
    { encoding: "utf8" }
  );
  const fileName = `frontity.settings.${fileExtension}`;
  const fileData = fileTemplate.replace(/\$([\w-]+)\$/g, (_match, key) => {
    if (key === "settings") return JSON.stringify(frontitySettings, null, 2);
  });
  await writeFile(fileName, fileData);
};

// This function installs the Frontity packages.
const installFrontityPackages = async ({ packages }: CreateOptions) => {
  const command = `npm install ${packages.join(" ")}`;
  await new Promise(resolve =>
    exec(command, { silent: true }, () => resolve())
  );
};

// This functions clones the starter theme.
const cloneStarterTheme = async ({ theme }: CreateOptions) => {
  const path = "packages/starter-theme";
  await ensureDir(path);
  process.chdir(path);
  const command = `npm pack ${theme}`;
  await new Promise(resolve =>
    exec(command, { silent: true }, () => resolve())
  );
  const tarball = (await readDir("./")).find(file => /\.tgz$/.test(file));
  await extract({ file: tarball, strip: 1 });
  await remove(tarball);
  process.chdir("../..");
};

// This function removes the files and directories created
// with `frontity create`.
const revertProgress = async ({ path }: CreateOptions) => {
  process.chdir(path);
  if (dirExisted) {
    const removableContent = [
      "node_modules",
      "frontity.settings.js",
      "frontity.settings.ts",
      "package.json",
      "package-lock.json",
      "packages"
    ];
    for (const content of removableContent) await remove(content);
  } else {
    await remove(path);
  }
};

export default async (passedOptions?: CreateOptions) => {
  // 1. Parses and validates options.
  const { emitter, ...options } = normalizeOptions(passedOptions);

  let step: Promise<void>;

  try {
    // 2. Ensures that the app dir exists and is empty.
    step = ensureAppDir(options);
    if (emitter) {
      emitter.emit(
        "create",
        `Ensuring ${chalk.yellow(options.path)} directory.`,
        step
      );
    }
    await step;

    // This nested try catch avoids removing the directory
    // if it existed before running frontity.
    try {
      // 3. Creates `package.json`.
      step = createPackageJson(options);
      if (emitter) {
        emitter.emit(
          "create",
          `Creating ${chalk.yellow("package.json")}`,
          step
        );
      }
      await step;

      // 4. Creates `frontity.settings`.
      const fileExtension = options.typescript ? "ts" : "js";
      step = createFrontitySettings(fileExtension, options);
      if (emitter) {
        emitter.emit(
          "create",
          `Creating ${chalk.yellow(`frontity.settings.${fileExtension}`)}`,
          step
        );
      }
      await step;

      // 5. Installs Frontity packages.
      step = installFrontityPackages(options);
      if (emitter) {
        emitter.emit(
          "create",
          `Installing ${options.packages
            .map(pkg => chalk.green(pkg))
            .join(", ")}`,
          step
        );
      }
      await step;

      // 6. Clones `@frontity/mars-theme` inside `packages`.
      step = cloneStarterTheme(options);
      if (emitter) {
        emitter.emit("create", `Cloning ${chalk.green(options.theme)}`, step);
      }
      await step;

      if (emitter) {
        emitter.emit(
          "create",
          `${chalk.bold(
            "\nFrontity project created"
          )}\n\nYou can start development with ${chalk.bold.green(
            "frontity dev"
          )}\nFor documentation visit ${chalk.underline.magenta(
            "https://docs.frontity.org/"
          )}\n`
        );
      }
    } catch (error) {
      await revertProgress(options);
      throw error;
    }
  } catch (error) {
    const message = `${chalk.bold.red("Error: ")}${chalk.red(error.message)}`;

    if (emitter) emitter.emit("create", message);
    else console.error(message);
  }
};
