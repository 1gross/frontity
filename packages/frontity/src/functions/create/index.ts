import chalk from "chalk";
import { Options } from "./types";
import {
  normalizeOptions,
  ensureProjectDir,
  createPackageJson,
  createFrontitySettings,
  cloneStarterTheme,
  installDependencies,
  revertProgress
} from "./steps";

const defaultOptions: Options = {
  path: process.cwd(),
  typescript: false,
  packages: ["frontity", "@frontity/file-settings"],
  theme: "@frontity/mars-theme"
};

export default async (passedOptions?: Options) => {
  // 1. Parses and validates options.
  const options = normalizeOptions(defaultOptions, passedOptions);

  // This functions will emit an event if an emitter is passed in options.
  const emit = (message: string, step?: Promise<void>) => {
    if (options.emitter) options.emitter.emit("create", message, step);
  };

  let step: Promise<any>;
  let dirExisted: boolean;

  try {
    // 2. Ensures that the project dir exists and is empty.
    step = ensureProjectDir(options);
    emit(`Ensuring ${chalk.yellow(options.path)} directory.`, step);
    dirExisted = await step;

    try {
      // This nested try catch avoids removing the directory
      // on error if it existed before running frontity.

      // 3. Creates `package.json`.
      step = createPackageJson(options);
      emit(`Creating ${chalk.yellow("package.json")}.`, step);
      await step;

      // 4. Creates `frontity.settings`.
      const extension = options.typescript ? "ts" : "js";
      step = createFrontitySettings(extension, options);
      emit(`Creating ${chalk.yellow(`frontity.settings.${extension}`)}.`, step);
      await step;

      // 5. Clones `@frontity/mars-theme` inside `packages`.
      step = cloneStarterTheme(options);
      emit(`Cloning ${chalk.green(options.theme)}.`, step);
      await step;

      // 6. Installs dependencies.
      step = installDependencies(options);
      emit(`Installing dependencies.`, step);
      await step;

      emit(
        `${chalk.bold(
          "\nFrontity project created."
        )}\n\nYou can start development with ${chalk.bold.green(
          "frontity dev"
        )}.\nFor documentation visit ${chalk.underline.magenta(
          "https://docs.frontity.org/"
        )}.\n`
      );
    } catch (error) {
      // Remove all files generated.
      await revertProgress(dirExisted, options);
      throw error;
    }
  } catch (error) {
    const message = `${chalk.bold.red("Error: ")}${chalk.red(error.message)}`;
    if (options.emitter) emit(message);
    else throw error;
  }
};
