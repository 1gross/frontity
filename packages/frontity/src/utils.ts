import { readdirSync } from "fs";

// This function checks if the path passed is an empty directory
// or an empty GitHub repository.
export const isEmptyDir = (path: string = process.cwd()): boolean => {
  try {
    // Get the content of the path.
    const dirContent = readdirSync(path);

    // Return true if the path is empty.
    if (!dirContent.length) return true;

    // Define a list of files/dirs allowed in an empty GitHub repo.
    const allowedContent = ["readme.md", "license", ".gitignore", ".git"];

    // Return true if all the content in path is allowed.
    if (
      !dirContent.filter(
        content => !allowedContent.includes(content.toLowerCase())
      ).length
    )
      return true;
  } catch (error) {
    // Throw if the path does not exist.
    if (error.code === "ENOENT")
      throw new Error("The path provided to `isEmptyDir` does not exist.");
  }

  // Return false if the path is not empty or does not contain an
  // empty GitHub repo.
  return false;
};

// This functions checks if the path passed is a Git repository.
export const isGitRepository = (path: string = process.cwd()): boolean => {
  try {
    // Gets the content of the path.
    const dirContent = readdirSync(path);

    // Returns true if the dir contains a `.git` folder.
    if (dirContent.includes(".git")) return true;
  } catch (error) {
    if (error.code === "ENOENT")
      throw new Error("The path provided to `isGitRepository` does not exist.");
  }

  // Returns false if the dir does not contain a `.git` folder.
  return false;
};
