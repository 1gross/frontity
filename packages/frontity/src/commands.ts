import program from "commander";
import { readFileSync } from "fs-extra";
import { create, createPackage, dev, build, serve } from "./actions";

const { version } = JSON.parse(
  readFileSync("./package.json", { encoding: "utf8" })
);

// Sets the version and the description of the program.
program.version(version).description("Frontity CLI");

// Registers a `create` command that takes an optional
// parameter called `name`. It also accepts the following
// options: --typescript, --use-cwd.
program
  .command("create [name]")
  .option("-t, --typescript", "Adds support for TypeScript")
  .option("-c, --use-cwd", "Generates the project in the current directory.")
  .description("Creates a new Frontity project.")
  .action(create);

program
  .command("create-package [name]")
  .description("Creates a new Frontity package.")
  .action(createPackage);

program
  .command("dev")
  .description("Starts the project in development mode.")
  .action(dev);

program
  .command("build")
  .description("Builds the project for production.")
  .action(build);

program
  .command("serve")
  .description("Starts a server in production mode.")
  .action(serve);

// Parses the parameters and adds them to the `command` object.
program.parse(process.argv);
