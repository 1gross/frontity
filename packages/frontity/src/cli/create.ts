import { resolve } from "path";
import ora from "ora";
import chalk from "chalk";
import { prompt, Question } from "inquirer";
import create from "../commands/create";
import { subscribe } from "../steps";
import { errorLogger, log } from "../utils";
import { Options } from "../steps/types";

export default async ({ name, typescript, useCwd }) => {
  const options: Options = {};

  if (!name) {
    const questions: Question[] = [
      {
        name: "name",
        type: "input",
        message: "Enter a name for the project:",
        default: "my-frontity-project"
      },
      {
        name: "theme",
        type: "input",
        message: "Enter a starter theme to clone:",
        default: "@frontity/mars-theme"
      }
    ];

    const answers = await prompt(questions);
    options.name = answers.name;
    options.theme = answers.theme;
    log();
  } else {
    options.name = name;
  }

  options.typescript = typescript;
  options.path = useCwd ? process.cwd() : resolve(process.cwd(), options.name);

  // Get the emitter for `create`
  const emitter = create(options);
  emitter.on("error", errorLogger);
  emitter.on("message", (message, action) => {
    if (action) ora.promise(action, message);
    else log(message);
  });

  await emitter;

  log(chalk.bold("\nFrontity project created.\n"));

  const subscribeQuestions: Question[] = [
    {
      name: "subscribe",
      type: "confirm",
      message: "Do you want to receive framework updates by email?",
      default: false
    },
    {
      name: "email",
      type: "input",
      message: "Please, enter your email:",
      when: answers => answers.subscribe
    }
  ];
  const answers = await prompt(subscribeQuestions);

  if (answers.subscribe) {
    log();

    emitter.on("subscribe", (message, action) => {
      if (action) ora.promise(action, message);
      else log(message);
    });

    await subscribe(answers.email);

    log("\nThanks for subscribing! ðŸ˜ƒ");
  } else {
    log(
      `\nOk, that's fine! ðŸ˜‰\nYou can subscribe at any point with ${chalk.bold.green(
        "npx frontity subscribe <email>"
      )}.`
    );
  }

  log(
    `\nRun ${chalk.bold.green(
      `cd ${options.name} && npx frontity dev`
    )} and have fun! ðŸŽ‰\n\nYou can find docs at ${chalk.underline.magenta(
      "https://docs.frontity.org/"
    )}.\nIf you have any doubts, join our community at ${chalk.underline.magenta(
      "https://community.frontity.org/"
    )}.\n`
  );
};
