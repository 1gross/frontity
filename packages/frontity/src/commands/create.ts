import { EOL } from "os";
import { resolve } from "path";
import {
  ensureDirSync,
  readdirSync as readDirSync,
  readFileSync,
  writeFileSync,
  removeSync
} from "fs-extra";
import { exec } from "shelljs";
import { extract } from "tar";
import { isEmptyDir, isGitRepository, normalizeOptions } from "../utils";
import { CreateOptions, PackageJson } from "../types";
import emitter from "../emitter";

const emit = (message: string) => emitter.emit("create", message);

export default (passedOptions?: CreateOptions) => {
  // 1. Parses and validates options.
  emit("Parsing and validating options");
  const { name, path, packages } = normalizeOptions(passedOptions);

  // 2. Ensures that the directory exists and is empty.
  {
    emit(`Creating project in ${path}`);
    ensureDirSync(path);
    process.chdir(path);
    if (!isEmptyDir(path))
      throw new Error("The directory passed to `create` function is not empty");
  }

  // 3. Initiates Git repository.
  {
    emit("Initiating Git repository");
    if (!isGitRepository(path)) {
      const command = "git init";
      exec(command, { silent: true });
    }
  }

  // 4. Generates `package.json`.
  {
    emit("Generating package.json");
    const packageJson: PackageJson = {
      name,
      version: "0.1.0",
      description: "Frontity project",
      keywords: ["frontity"],
      scripts: {
        dev: "frontity dev",
        build: "frontity build",
        serve: "frontity serve"
      }
    };
    const fileName = "package.json";
    const fileData = `${JSON.stringify(packageJson, null, 2)}${EOL}`;
    writeFileSync(fileName, fileData);
  }

  // 5. Generates `frontity.settings.js`.
  {
    emit("Generating frontity.settings.js");
    const frontitySettings = { name, packages };
    const fileName = "frontity.settings.js";
    const fileData = `module.exports = ${JSON.stringify(
      frontitySettings,
      null,
      2
    )}${EOL}`;
    writeFileSync(fileName, fileData);
  }

  // 6. Generates `.gitignore`.
  {
    emit("Generating .gitignore");
    const gitIgnore = readFileSync(
      resolve(__dirname, "../templates/gitignore-template"),
      { encoding: "utf8" }
    );
    const fileName = ".gitignore";
    const fileData = `${gitIgnore}${EOL}`;
    writeFileSync(fileName, fileData);
  }

  // 7. Generates `README.md`.
  {
    emit("Generating README.md");
    const readme = readFileSync(
      resolve(__dirname, "../templates/readme-template"),
      { encoding: "utf8" }
    );
    const fileName = "README.md";
    const fileData = `${readme.replace(/\$(\w+)\$/g, (_match, key) => {
      if (key === "project_name") return name;
    })}${EOL}`;
    writeFileSync(fileName, fileData);
  }

  // 8. Installs Frontity packages.
  {
    emit(`Installing ${packages.join(", ")}`);
    const command = `npm i ${packages.join(" ")}`;
    exec(command, { silent: true });
  }

  // 9. Clones `@frontity/starter-theme` inside `packages`.
  {
    emit("Cloning @frontity/starter-theme");
    const path = "packages/starter-theme";
    ensureDirSync(path);
    process.chdir(path);
    const command = "npm pack @frontity/h2r";
    exec(command, { silent: true });
    const tarball = readDirSync("./").find(file => /\.tgz$/.test(file));
    extract({ file: tarball, sync: true, strip: 1 });
    removeSync(tarball);
    process.chdir("../..");
  }

  // 10. Commit changes to Git.
  {
    emit("Commiting changes");
    const command = 'git add . && git commit -m "Setup Frontity"';
    exec(command, { silent: true });
  }
};
