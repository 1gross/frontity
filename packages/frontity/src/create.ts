import { isEmptyDir } from "./utils";

type Options = {
  // Name of the app.
  name: string;
  // Path where the app should be created.
  path?: string;
  // Generate app inside a new directory.
  cwd?: boolean;
  // Support for TypeScript.
  typescript?: boolean;
  // Frontity packages that need to be installed.
  packages?: string[];
};

// There are different options of how this function should work:
// 1. We need a new dir so we create a new dir named after `options.name`
//    inside the path `options.path`.
// 2. We don't need a new dir, so we generate the app in the cwd or path passed.

export default (options: Options) => {
  isEmptyDir(process.cwd());
};
