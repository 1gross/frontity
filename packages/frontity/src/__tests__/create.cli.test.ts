import { resolve } from "path";

import { EventPromised } from "../utils/eventPromised";
import * as createCmd from "../commands/create";
import create from "../cli/create";
import * as inquirer from "inquirer";
import * as utils from "../utils";

jest.mock("../commands/create");
jest.mock("inquirer");

const mockedCreateCmd = createCmd as jest.Mocked<typeof createCmd>;
const mockedInquirer = inquirer as jest.Mocked<typeof inquirer>;
const mockedUtils = utils as any;

describe("CLI create", () => {
  beforeEach(() => {
    mockedCreateCmd.default.mockReset();
    mockedCreateCmd.default.mockReturnValue(
      new EventPromised(resolve => resolve())
    );
    mockedInquirer.prompt.mockReset();
    mockedInquirer.prompt.mockResolvedValue({
      name: "test-project",
      theme: "test-theme"
    });
    mockedUtils.errorLogger = jest.fn();
  });

  test("frontity create", async done => {
    const options = {
      name: undefined,
      typescript: undefined,
      useCwd: undefined,
      prompt: true
    };
    await create(options);

    expect(mockedInquirer.prompt.mock.calls).toMatchSnapshot();
    expect(mockedInquirer.prompt).toHaveBeenCalledTimes(2);
    expect(mockedInquirer.prompt.mock.calls[0][0]).toMatchObject([
      { message: "Enter a name for the project:" },
      { message: "Enter a starter theme to clone:" }
    ]);

    expect(mockedInquirer.prompt.mock.calls[1][0]).toMatchObject([
      { message: "Do you want to receive framework updates by email?" },
      { message: "Please, enter your email:" }
    ]);
    done();
  });

  test("frontity create 'test-project'", () => {
    const options = {
      name: "test-project",
      typescript: undefined,
      useCwd: undefined,
      prompt: true
    };

    // Don't need to await because we only check the args
    // that the create command was called with
    create(options);
    expect(mockedCreateCmd.default.mock.calls).toMatchSnapshot();
    expect(mockedCreateCmd.default).toHaveBeenCalledWith({
      name: options.name,
      typescript: false,
      path: resolve(process.cwd(), options.name)
    });
  });

  test("frontity create 'test-project' --typescript", () => {
    const options = {
      name: "test-project",
      typescript: true,
      useCwd: undefined,
      prompt: true
    };

    // Don't need to await because we only check the args
    // that the create command was called with
    create(options);
    expect(mockedCreateCmd.default.mock.calls).toMatchSnapshot();
    expect(mockedCreateCmd.default).toHaveBeenCalledWith({
      name: options.name,
      typescript: true,
      path: resolve(process.cwd(), options.name)
    });
  });

  test("frontity create --no-prompt", async () => {
    const options = {
      name: undefined,
      typescript: true,
      useCwd: undefined,
      prompt: false
    };

    try {
      await create(options);
    } catch (err) {
      expect(err.message).toBe("You need to provide the name for the project");
    }
  });

  test("FRONTITY_NAME='test-project'; frontity create --no-prompt", () => {
    const name = "test-project";
    process.env.FRONTITY_NAME = name;

    const options = {
      name: undefined,
      typescript: undefined,
      useCwd: undefined,
      prompt: false
    };

    // Don't need to await because we only check the args
    // that the create command was called with
    create(options);
    expect(mockedCreateCmd.default.mock.calls).toMatchSnapshot();
    expect(mockedCreateCmd.default).toHaveBeenCalledWith({
      name,
      typescript: false,
      path: resolve(process.cwd(), name)
    });
  });

  test("FRONTITY_TYPESCRIPT='true'; frontity create 'test-project' --no-prompt", () => {
    process.env.FRONTITY_TYPESCRIPT = "true";

    const options = {
      name: "test-project",
      typescript: undefined,
      useCwd: undefined,
      prompt: false
    };

    // Don't need to await because we only check the args
    // that the create command was called with
    create(options);
    expect(mockedCreateCmd.default.mock.calls).toMatchSnapshot();
    expect(mockedCreateCmd.default).toHaveBeenCalledWith({
      name: options.name,
      typescript: true,
      path: resolve(process.cwd(), options.name)
    });
  });
});
