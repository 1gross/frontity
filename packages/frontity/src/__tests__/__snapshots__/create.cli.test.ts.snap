// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI create FRONTITY_NAME='test-project'; frontity create --no-prompt 1`] = `"test-project"`;

exports[`CLI create FRONTITY_NAME='test-project'; frontity create --no-prompt 2`] = `false`;

exports[`CLI create FRONTITY_TYPESCRIPT='true'; frontity create 'test-project' --no-prompt 1`] = `"test-project"`;

exports[`CLI create FRONTITY_TYPESCRIPT='true'; frontity create 'test-project' --no-prompt 2`] = `true`;

exports[`CLI create frontity create 'test-project' --typescript 1`] = `"test-project"`;

exports[`CLI create frontity create 'test-project' --typescript 2`] = `true`;

exports[`CLI create frontity create 'test-project' 1`] = `"test-project"`;

exports[`CLI create frontity create 'test-project' 2`] = `false`;

exports[`CLI create frontity create 1`] = `
Array [
  Array [
    Array [
      Object {
        "default": "my-frontity-project",
        "message": "Enter a name for the project:",
        "name": "name",
        "type": "input",
      },
      Object {
        "default": "@frontity/mars-theme",
        "message": "Enter a starter theme to clone:",
        "name": "theme",
        "type": "input",
      },
    ],
  ],
  Array [
    Array [
      Object {
        "default": false,
        "message": "Do you want to receive framework updates by email?",
        "name": "subscribe",
        "type": "confirm",
      },
      Object {
        "message": "Please, enter your email:",
        "name": "email",
        "type": "input",
        "when": [Function],
      },
    ],
  ],
]
`;
